### DISCUSSION TITLE
no changes added to commit when doing environment specific changes

### DISCUSSION THREAD

**Discussion by hghtwr** [akuity/kargo#2967] [Q&A] [answered]
Hey all, 

trying around with kargo a lot at the moment and I encountered a question about it.

I have the following setup for my source repo: 

```
app/
├─ charts/
    ├─ templates/
    ├─ values-dev.yaml
    ├─ values-uat.yaml
    ├─ values-prod.yaml
├─ stages/
    ├─ dev/
         ├─ manifests.yaml
    ├─ uat/
    ├─ prod/
```

I created stages that subscribe to each other in the following order: 
- dev
- uat
- prod

So now the problem: 
1. Promote `queenly-alligator` from dev to uat. 
2. Notice that you made a mistake in the host url in the `values-uat.yaml`.
3. Change `values-uat.yaml` -> It creates the freight `warped-ostrich`
4. Try to promote `warped-ostrich` to dev.
5. Promote to dev fails because: git commit - no changes added to commit

So in short: If I only change an environment specific config in my helm chart, I will not be able to commit to the previous stages as there are not changes.

I understand I can make a manual promotion (and it works) but I was wondering if there's a better way to handle this which I might have missed. Or maybe I'm using the wrong approach here.

Thx!

**Comment by krancour** [comment]
This _could_ be a bug in the git commit step and I will look into it today.

In the legacy promotion mechanisms that were retired when we went GA, git-based promotions were tolerant of promotions with no diffs that resulted in no commit. The git commit step is supposed to work the same way, but it seems perhaps that it is not.

I'll look into it.

**Reply by hghtwr** [reply]
Great, already good to know for my concepts that there should be a way to tolerate commits without changes.
Thanks!

**Reply by krancour** [reply]
This is perplexing. What you are encountering seems as if it should not be possible:

https://github.com/akuity/kargo/blob/947e4f0c9eb7297920b4f40b0f206847cdf7fa83/internal/directives/git_commiter.go#L81-L106

**Reply by krancour** [reply]
If there are no diffs from the head, it just returns the commit ID of the head as its state is identical to what you _would_ have committed.

**Reply by hghtwr** [reply]
Is there a good way for me to debug what's going on in this PromotionStep during it's execution?

**Reply by krancour** [reply]
Not a convenient one.

Can you please share the details of your promotion process? i.e. Show the sequence of steps you're using?

**Reply by hghtwr** [reply]
So, this is my promotion process. I try to merge back into the same branch but a different folder.

```
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: teststable-tags
  namespace: kargo-poc-tags
spec:
  requestedFreight:
  - origin:
      kind: Warehouse
      name: kargo-poc-tags
    sources:
      direct: true
  verification:
    analysisTemplates:
    - name: kargo-poc-teststable
  promotionTemplate:
    spec:
      steps:
      - uses: git-clone
        config:
          repoURL: https://<...>-service-cicd.git
          checkout:
          - branch: develop
            path: ./dir
      - uses: git-clear
        config: 
          path: ./dir/stages/teststable
      - uses: helm-update-image
        as: update-image
        config:
          path: ./dir/chart/techcelerate-service/values-teststable.yaml
          images:
          - image: <...>-service-cicd-service
            key: image.tag
            value: Tag
      - uses: helm-template
        config:
          releaseName: techcelerate-service
          path: ./dir/chart/techcelerate-service
          valuesFiles: 
            - ./dir/chart/techcelerate-service/values-teststable.yaml
          outPath: ./dir/stages/teststable/manifests.yaml
      - uses: git-commit
        as: commit
        config:
          path: ./dir/stages/teststable
          messageFromSteps:
          - update-image
      - uses: git-push
        as: push
        config:
          path: ./dir/stages/teststable
          generateTargetBranch: true
      - uses: git-open-pr
        as: open-pr
        config: 
          repoURL: https://<...>-service-cicd.git
          sourceBranchFromStep: push
          targetBranch: develop
      - uses: git-wait-for-pr
        as: wait-pr
        config: 
          repoURL: https://<...>-service-cicd.git
          provider: gitlab
          prNumberFromStep: open-pr
      - uses : argocd-update
        config: 
          apps: 
            - name: techcelerate-service-teststable
              namespace: argocd
              sources: 
                - repoURL: https://<...>-service-cicd.git
                  desiredCommitFromStep: wait-pr
```

**Reply by krancour** [reply, marked as answer]
We assume the `path` for steps like `git-commit` or `git-push` to be the path to the _root_ of working tree, but you have provided the path to a directory _within_ a working tree.

I have a suspicion that defying this assumption is indirectly the cause of some unanticipated behavior. I need to put this theory to the test to see if this actually _is_ a problem. If it is, we'll either eliminate that assumption or validate it before proceeding.

We're working hard right now to release v1.1 before the (American) Thanksgiving holiday, so this may have to wait until after then.

In the meantime, you could perhaps try pointing at the root of the working tree instead. I suspect you've done it the way you have to avoid committing the changes that `helm-update-image` made to ` ./dir/chart/techcelerate-service/values-teststable.yaml`. An alternative workaround would be to use the `copy` step to copy that file _outside_ the working tree, use `helm-update-image` on the copy, and again use the copy in the `helm-template` step. Then you could safely use `./dir` as the path for the `git-commit` and `git-push` steps.

Hopefully the above will get you moving again and if it does, it helps prove my theory. (We'd still take more strategic action on it after the holiday.)

**Reply by hghtwr** [reply]
The copy did the trick :) Now it creates a MR without any changes. Indeed, it outputs the previous commit id.
I basically copied the sources outside of the working tree, did the helm update and template there and then used the working tree as `outPath` for the `helm-template`.

This solves my problem. 

**Thank you** so much for this speedy support, it's awesome!
I think you have built a pretty good product that closes a gap we worried about quite some time.

**Reply by krancour** [reply]
Thanks for the kind words and also for verifying my suspicion that path _must_ point at the root of the working tree. I'll open an issue and make sure we do something about that.

**Reply by krancour** [reply]
@hghtwr I figured out exactly what happened with your initial workflow:

1. `./dir/stages/teststable/manifests.yaml` was already what it was supposed to look like.
1. `./dir/chart/techcelerate-service/values-teststable.yaml` got updated.
1. When you committed, using (unanticipated) non-root path `./dir/stages/teststable`:
    1. A diff _was_ found. It would have been `../../chart/techcelerate-service/values-teststable.yaml`. This would have allowed a commit to proceed.
    1. `.` (`./dir/stages/teststable`)  would than have been staged for commit, and _that_ path did not contain any diffs.

**Reply by krancour** [reply]
I opened #2983 to fix this.
